OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java

```diff
@@ -1742,48 +1747,69 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
 		} else {
 			return null;
 		}
+
 		int currentLanesCount = countLanesMinOne(currentSegm);
 		int prevLanesCount = countLanesMinOne(prevSegm);
 		boolean oneLane = currentLanesCount == 1 && prevLanesCount == 1;
 		int[] lanes;
+
 		if (oneLane) {
 			lanes = createCombinedTurnTypeForSingleLane(rs, deviation);
 		} else {
-			int ls = currentLanesCount + rs.leftLanes + rs.rightLanes;
-			lanes = new int[ls];
-			for (int it = 0; it < ls; it++) {
-				if (it < rs.leftLanes) {
-					// lanes left from active
-					if (laneType != TurnType.C) {
-						// avoid repeat of turns, e.g. for TSLL get TL
-						lanes[it] = TurnType.getPrev(laneType) << 1;
-					} else {
-						// can be several straight directions
-						lanes[it] = TurnType.C << 1;
-					}
-				} else if (it >= rs.leftLanes + currentLanesCount) {
-					// lanes in right from active
-					if (laneType != TurnType.C) {
-						lanes[it] = TurnType.getNext(laneType) << 1;
-					} else {
-						lanes[it] = TurnType.C << 1;
+			lanes = new int[prevLanesCount];
+			boolean ltr = rs.leftLanes == rs.rightLanes ? rs.roadsOnLeft < rs.roadsOnRight
+					: rs.leftLanes < rs.rightLanes;
+			// active lanes
+			for(int i = 0; i < Math.min(lanes.length, currentLanesCount); i++) {
+				int ind = ltr ? i : lanes.length - i - 1;
+				lanes[ind] = (laneType << 1) + 1;
+			}
+			// left lanes
+			for(int i = 0; i < Math.min(lanes.length, rs.leftLanes); i++) {
+				int lane = getTurnByAngle(rs.attachedAngles.get(i));
+				if (lanes[i] == 0) {
+					lanes[i] = lane << 1;
```

Comment:
If laneType equal e.g. TSLL, then left lane must be TL. Otherwise can be repeat active/inactive lanes that will be misunderstood users.
Therefore in previous code was row:
lanes[it] = TurnType.getPrev(laneType) << 1;

Reply:
Ready to check